### PR TITLE
Deriving cltv and example

### PR DESCRIPTION
**Pull request by rustyrussell**
Sooo...many...words...

### CODE REVIEW

File: 07-routing-gossip.md [hunk 1]
```diff
@@ -441,6 +448,93 @@ Other more advanced considerations involve diversity of routes to
...
+If B were to send 4,999,999 millisatoshi to C, it wouldn't charge
+itself a fee, but it would need to consider the `cltv_expiry_delta` in
+the B->C `channel_update` (ie. the `cltv_expiry_delta` C told B in
+`open_channel` or `accept_channel`).  We also assume it adds a "shadow
+route" to give an extra CLTV of 42.
+
+   * `amount_msat`: 4999999
+   * `cltv_expiry`: current-block-height + 30 + 42
+   * `onion_routing_packet`:
+	 * `amt_to_forward` = 4999999
```
Comment (cdecker) [inline 07-routing-gossip.md]: nit: tab misaligns the source

File: 02-peer-protocol.md [hunk 2]
```diff
@@ -114,6 +115,9 @@ The `temporary_channel_id` is used to identify this channel until the funding tr
 
 `feerate_per_kw` indicates the initial fee rate by 1000-weight (ie. 1/4 the more normally-used 'feerate per kilobyte') which this side will pay for commitment and HTLC transactions as described in [BOLT #3](03-transactions.md#fee-calculation) (this can be adjusted later with an `update_fee` message).  `to_self_delay` is the number of blocks that the other nodes to-self outputs must be delayed, using `OP_CHECKSEQUENCEVERIFY` delays; this is how long it will have to wait in case of breakdown before redeeming its own funds.
 
+`cltv_expiry_delta` indicates the CLTV timeout requirement for HTLCs
+offered into this channel.
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md]: Well, it also covers the terminal case, so the language here is deliberately vague.

File: 02-peer-protocol.md [hunk 3]
```diff
@@ -141,6 +145,8 @@ compressed secp256k1 pubkeys. The sender SHOULD set `feerate_per_kw`
 to at least the rate it estimates would cause the transaction to be
 immediately included in a block.
 
+The sender MUST set `cltv_expiry_delta` to the number of blocks it will
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md]: Good point.  It's certainly easy to relax this.

File: 02-peer-protocol.md [hunk 4]
```diff
@@ -508,7 +515,7 @@ the incoming HTLC has been irrevocably committed.
 A node MUST NOT fail an incoming HTLC (`update_fail_htlc`) for which it has committed 
 to an outgoing HTLC, until the removal of the outgoing HTLC is irrevocably committed, or the outgoing on-chain HTLC output has been spent via the HTLC-timeout transaction with sufficient depth.
 
-A node MUST fail an incoming HTLC (`update_fail_htlc`) once its `cltv_expiry` has been reached.
+A node MUST fail an incoming HTLC (`update_fail_htlc`) once its `cltv_expiry` has been reached, or if its `ctlv_expiry` is within `cltv_expiry_delta` (as this node sent in `open_channel` or `accept_channel`) of the current block height.
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md]: Ack.

File: 02-peer-protocol.md [hunk 5]
```diff
@@ -520,61 +527,113 @@ Fulfilling an HTLC is different: knowledge of the preimage is by definition irre
...
-commitment transaction it has with C and spending the HTLC output.
-Even though C might race to try to use its payment preimage at that point to
-also spend the HTLC, it should be resolved well before the 3 day
-deadline so B can either redeem the HTLC off A or close it.
+Consider the following scenario, where A sends an HTLC to B, who
+forwards to C, who delivers the goods as soon as the payment is
+received.
+
+1.  C needs to be sure that the HTLC from B cannot time out, even if B becomes
+    unresponsive; ie. C can fulfill the incoming HTLC on-chain before B can
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md]: Will add sp. delta to fix the others, too :)

File: 02-peer-protocol.md [hunk 6]
```diff
@@ -520,61 +527,113 @@ Fulfilling an HTLC is different: knowledge of the preimage is by definition irre
...
+Consider the following scenario, where A sends an HTLC to B, who
+forwards to C, who delivers the goods as soon as the payment is
+received.
+
+1.  C needs to be sure that the HTLC from B cannot time out, even if B becomes
+    unresponsive; i.e. C can fulfill the incoming HTLC on-chain before B can
+    time it out on-chain.
+
+2.  B needs to be sure that if C fulfills the HTLC from B, it can fulfill the
+    HTLC from A.  i.e. B can get the preimage from C and fulfill incoming the
```
Comment (jimpo) [inline 02-peer-protocol.md]: fulfill the incoming

File: 02-peer-protocol.md [hunk 7]
```diff
@@ -520,61 +527,113 @@ Fulfilling an HTLC is different: knowledge of the preimage is by definition irre
...
+    HTLC on-chain before B can time it out on-chain.
```
Comment (jimpo) [inline 02-peer-protocol.md]: before A can time it out on-chain

File: 02-peer-protocol.md [hunk 8]
```diff
@@ -520,61 +527,113 @@ Fulfilling an HTLC is different: knowledge of the preimage is by definition irre
...
+  transaction being included in a block.
+
+The worst case is for a forwarding node (B) which takes the longest
+possible time to spot the outgoing HTLC fulfillment, and then takes
+the longest possible time to redeem it on-chain:
+
+1. The B->C HTLC times out at block `N`, and B waits `G` blocks until
+   it gives up waiting for C.  B or C commits to the blockchain,
+   and B spends HTLC.
+2. Bad case: C wins the race and fulfills the HTLC, B only sees
```
Comment (jimpo) [inline 02-peer-protocol.md]: What if neither B nor C's block makes it into the first block after `G`? Should this take into account `S`, B's time to confirm? (or C's in the case that C wins the race at the latest possible block)

File: 07-routing-gossip.md [hunk 9]
```diff
@@ -441,6 +448,95 @@ Other more advanced considerations involve diversity of routes to
...
+The `update_add_htlc` from B to C would be the same as the B->C direct
+payment above.
+
+Finally, if for some reason A chose the more expensive route via D, it
+would send the following `update_add_htlc` to D:
+
+   * `amount_msat`: 5020398
+   * `cltv_expiry`: current-block-height + 70 + 42
+   * `onion_routing_packet`:
+	 * `amt_to_forward` = 4999999
```
Comment (jimpo) [inline 07-routing-gossip.md]: nit: indentation

File: 02-peer-protocol.md (hunk 6 above)
Comment (rustyrussell) [reply 02-peer-protocol.md]: OK.

File: 02-peer-protocol.md (hunk 7 above)
Comment (rustyrussell) [reply 02-peer-protocol.md]: Ack, good catch.

File: 02-peer-protocol.md (hunk 8 above)
Comment (rustyrussell) [reply 02-peer-protocol.md]: Conservatively, I think you're right.  B will spend via time out (worst case) at N+G+S+R.  So C can fulfill at any time <= N+G+S+R.

That gives us 3R+2G+2S; that changes our recommendation from 9 to 12.

However, if @Roasbeef wants a higher S value, then it increases again (as would our recommended final cltv value).

File: 07-routing-gossip.md (hunk 9 above)
Comment (rustyrussell) [reply 07-routing-gossip.md]: It's nested correctly AFAICT?

**Comment by rustyrussell** [comment]
OK, the example works, FWIW.

File: 02-peer-protocol.md [hunk 10]
```diff
@@ -508,7 +508,7 @@ the incoming HTLC has been irrevocably committed.
 A node MUST NOT fail an incoming HTLC (`update_fail_htlc`) for which it has committed 
 to an outgoing HTLC, until the removal of the outgoing HTLC is irrevocably committed, or the outgoing on-chain HTLC output has been spent via the HTLC-timeout transaction with sufficient depth.
 
-A node MUST fail an incoming HTLC (`update_fail_htlc`) once its `cltv_expiry` has been reached.
+A node MUST fail an incoming HTLC (`update_fail_htlc`) once its `cltv_expiry` has been reached, or if `current_height` - `cltv_expiry` < `cltv_expiry_delta` for the outgoing channel.
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md]: Yep!  Fixed.

File: 02-peer-protocol.md [hunk 11]
```diff
@@ -520,61 +520,113 @@ Fulfilling an HTLC is different: knowledge of the preimage is by definition irre
...
+the longest possible time to redeem it on-chain:
+
+1. The B->C HTLC times out at block `N`, and B waits `G` blocks until
+   it gives up waiting for C.  B or C commits to the blockchain,
+   and B spends HTLC, which takes `S` blocks to be included.
+2. Bad case: C wins the race (just) and fulfills the HTLC, B only sees
+   that transaction when it sees block `N+G+S+1`.
+3. Worst case: There's reorganization `R` deep in which C wins and
+   fulfills. B only sees transaction at `N+G+S+R`.
+4. B now needs to fulfill the incoming A->B HTLC, but waits `G` more
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md]: A hasn't done anything wrong (yet); it was C that forced B<->C onchain. So B tries to fulfill, but A is unresponsive...

File: 07-routing-gossip.md (hunk 9 above)
Comment (jimpo) [reply 07-routing-gossip.md]: It's indented with a tab instead of spaces, which makes it appear misaligned with the next field on GitHub.

**Comment by rustyrussell** [comment]
OK, I've assigned to @Roasbeef who said he wanted to review calcs; if happy, please commit.  Will do rebase and squash (damn spelling fixes) now.

**Pull request merged**